pallets/asset/src/identifier.rs

```diff
@@ -0,0 +1,203 @@
+use codec::{Decode, Encode};
+use core::convert::TryInto;
+use sp_std::vec;
+
+/// Implementation of common asset identifiers
+/// https://www.cusip.com/identifiers.html
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum Identifier {
+    CUSIP([u8; 9]),
+    CINS([u8; 9]),
+    ISIN([u8; 12]),
+    LEI([u8; 20]),
+    EMPTY,
+}
+
+impl Default for Identifier {
+    fn default() -> Self {
+        Identifier::EMPTY
+    }
+}
+
+impl Identifier {
+    /// Validate `bytes` is a valid CUSIP identifier, returns an instance of `Identifier` if successful
+    pub fn cusip(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CUSIP(bytes))
+    }
+
+    /// Validate `bytes` is a valid CINS identifier, returns an instance of `Identifier` if successful
+    pub fn cins(bytes: [u8; 9]) -> Option<Identifier> {
+        Self::cusip(bytes).map(|_| Identifier::CINS(bytes))
+    }
+
+    /// Validate `bytes` is a valid ISIN identifier, returns an instance of `Identifier` if successful
+    pub fn isin(bytes: [u8; 12]) -> Option<Identifier> {
+        let (s1, s2) = bytes
+            .iter()
+            .copied()
+            .map(byte_value)
+            .flat_map(|b| if b > 9 { vec![b / 10, b % 10] } else { vec![b] })
+            .rev()
+            .enumerate()
+            .fold((0, 0), |(mut s1, mut s2), (i, digit)| {
+                if i % 2 == 0 {
+                    s1 += digit;
+                } else {
+                    s2 += 2 * digit;
+                    if digit >= 5 {
+                        s2 -= 9;
+                    }
+                }
+                (s1, s2)
+            });
+        ((s1 + s2) % 10 == 0).then_some(Identifier::ISIN(bytes))
+    }
+
+    /// Validate `bytes` is a valid LEI identifier, returns an instance of `Identifier` if successful
+    pub fn lei(bytes: [u8; 20]) -> Option<Identifier> {
+        bytes[..18]
+            .try_into()
+            .ok()
+            .map(lei_checksum)
+            .filter(|hash| *hash == (bytes[18] - b'0') * 10 + (bytes[19] - b'0'))
+            .map(|_| Identifier::LEI(bytes))
+    }
+
+    /// Ensures the identifier is valid.
+    /// Mainly used for validating manual constructions of the enum (user input).
+    pub fn validate(self) -> Option<Self> {
+        match self {
+            Identifier::CUSIP(bytes) => Self::cusip(bytes),
+            Identifier::CINS(bytes) => Self::cins(bytes),
+            Identifier::ISIN(bytes) => Self::isin(bytes),
+            Identifier::LEI(bytes) => Self::lei(bytes),
+            Identifier::EMPTY => Some(Identifier::EMPTY),
```

Comment:
Let's return a `None` here. The user should not able to submit `Identifier::EMPTY`.